File: Object/Brdf.cpp
#include "Object/Brdf.hpp"

#include <vector>

namespace Object {
    Brdf::~Brdf()
    {
    }

    Math::Color Brdf::reflected(const Math::Vector &, const Math::Normal &, const Math::Vector &, const Math::Color &) const
    {
        return Math::Color(0, 0, 0);
    }

    Math::Color Brdf::transmitted(const Math::Vector &, const Math::Normal &, const Math::Color &) const
    {
        return Math::Color(0, 0, 0);
    }

    float Brdf::lambert() const
    {
        return 0;
    }

    Math::Vector Brdf::sample(Math::Sampler &, const Math::Normal &, const Math::Vector &) const
    {
        return Math::Vector(0, 0, 0);
    }

    float Brdf::pdf(const Math::Vector &, const Math::Normal &, const Math::Vector &) const
    {
        return 0;
    }

    bool Brdf::opaque() const
    {
        return true;
    }

    float Brdf::transmitIor() const
    {
        return 1;
    }
}

File: src/object/brdf.rs
mod lambert;
pub use lambert::Lambert;

mod oren_nayar;
pub use oren_nayar::OrenNayar;

mod phong;
pub use phong::Phong;

mod torrance_sparrow;
pub use torrance_sparrow::TorranceSparrow;

use crate::geo;
use geo::Normal3;
use geo::Vec3;

use crate::object;
use object::Color;
use object::Sampler;

pub trait Brdf: Send + Sync {
    fn reflected(&self, dir_in: Vec3, nrm: Normal3, dir_out: Vec3, albedo: Color) -> Color;
    fn transmitted(&self, dir_in: Vec3, nrm: Normal3, albedo: Color) -> Color;
    fn lambert(&self) -> f32;

    fn sample(&self, sampler: &mut dyn Sampler, nrm: Normal3, dir_out: Vec3) -> Vec3;
    fn pdf(&self, dir_in: Vec3, nrm: Normal3, dir_out: Vec3) -> f32;

    fn opaque(&self) -> bool {
        return true;
    }
}
